src/tcp.rs

```diff
@@ -50,6 +53,16 @@ pub fn configure_socket(socket: SockRef, so_mark: &Option<u32>) -> Result<(), an
     Ok(())
 }
 
+// Interweave v4 and v6 addresses as per RFC8305.
+// The first address is v6 if we have any v6 addresses.
+fn sort_socket_addrs(socket_addrs: Vec<SocketAddr>) -> impl Iterator<Item = SocketAddr> {
+    let (addrs_v4, addrs_v6): (Vec<_>, Vec<_>) = socket_addrs.into_iter().partition(|a| match a {
+        SocketAddr::V4(_) => true,
+        SocketAddr::V6(_) => false,
+    });
+    addrs_v6.into_iter().interleave(addrs_v4)
```

Comment:
I can implement this manually if you don't prefer to add the itertools dependency.

Reply:
It would be preferable, yes, if I can avoid 1 more dependency.
Something like that is enough

```rust
// Interweave v4 and v6 addresses as per RFC8305.
// The first address is v6 if we have any v6 addresses.
fn sort_socket_addrs(socket_addrs: &[SocketAddr]) -> impl Iterator<Item = &SocketAddr> {
    let mut pick_v6 = true;
    let mut v6 = socket_addrs.iter().filter(|s| matches!(s, SocketAddr::V6(_)));
    let mut v4 = socket_addrs.iter().filter(|s| matches!(s, SocketAddr::V4(_)));
    std::iter::from_fn(move || {
        if pick_v6 {
            pick_v6 = false;
            v6.next().or(v4.next())
        } else {
            pick_v6 = true;
            v4.next().or(v6.next())
        }
    })
}
```

Reply:
the for loop turn into, note the copied()
```
 for (ix, addr) in sort_socket_addrs(&socket_addrs).copied().enumerate() {
```

Reply:
Done!